src/liballoc/task.rs

```diff
@@ -18,24 +18,70 @@ use crate::sync::Arc;
 /// exists as an alternative for those systems.
 ///
 /// [arc]: ../../std/sync/struct.Arc.html
-#[unstable(feature = "wake_trait", issue = "69912")]
+///
+/// # Examples
+///
+/// ```rust
+/// use std::future::Future;
+/// use std::sync::Arc;
+/// use std::task::{Context, Poll, Wake};
+/// use std::thread::{self, Thread};
+///
+/// /// A waker that wakes up the current thread when called.
+/// struct ThreadWaker(Thread);
+///
+/// impl Wake for ThreadWaker {
+///     fn wake(self: Arc<Self>) {
+///         self.0.unpark();
+///     }
+/// }
+///
+/// /// Run a future to completion on the current thread.
+/// ///
+/// /// This function may deadlock in nested calls; it should
+/// /// only be called once at the root of fn main.
+/// fn block_on<T>(fut: impl Future<Output = T>) -> T {
+///     // Pin the future so it can be polled.
+///     let mut fut = Box::pin(fut);
+///
+///     // Create a new context to be passed to the future.
+///     let t = thread::current();
+///     let waker = Arc::new(ThreadWaker(t)).into();
+///     let mut cx = Context::from_waker(&waker);
+///
+///     // Run the future to completion.
+///     loop {
+///         match fut.as_mut().poll(&mut cx) {
+///             Poll::Ready(res) => return res,
+///             Poll::Pending => thread::park(),
+///         }
+///     }
+/// }
```

Comment:
The example is really nice touch, demonstrating the real world use case without being overly complex.

The implementation contains one subtle issue: the wakeups might be lost if future itself uses thread parking & unparking mechanism. If fixing this would make the example too complex it probably should come with a warning. Currently there is no block_on in the std so it doesn't seem to far fetched that someone would copy and use this code as is.

Reply:
`thread::park` can also have spurious wakeups, so I think additional logic is needed here (maybe an `AtomicBool` that gets set by the waker?)

Reply:
Ah yeah; the right solution would be to use something like the [`parking`](https://docs.rs/parking/1.0.5/parking/index.html) crate. Given we don't have something like this in the stdlib I'm thinking we should probably issue a warning that this is not an implementation suitable for real-world use and merely for educational purposes.

Reply:
I'm adding a warning to the example to emphasize that this is not intended to be used in production:

> **Note:** This example trades correctness for simplicity. In order to prevent
> deadlocks, production-grade implementations will also need to handle
> intermediate calls to `thread::unpark` as well as nested invocations.

That way we can preserve the simplicity of the example, while discouraging copy-pasting the code for production use.